modules/ipc/include/hephaestus/ipc/zenoh/service.h

```diff
@@ -4,30 +4,146 @@
+  } else {
+    RequestT request;
+    auto payload = query.get_value().get_payload();
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+    serdes::deserialize<RequestT>(buffer, request);
+    return request;
+  }
+}
+
+template <typename RequestT, typename ReplyT>
+Service<RequestT, ReplyT>::Service(SessionPtr session, TopicConfig topic_config, Callback&& callback)
+  : session_(std::move(session)), topic_config_(std::move(topic_config)), callback_(std::move(callback)) {
+  checkTemplatedTypes<RequestT, ReplyT>();
+
+  auto query = [this](const zenohc::Query& query) mutable {
+    DLOG(INFO) << fmt::format("Received query from '{}'", query.get_keyexpr().as_string_view());
+
+    auto reply = this->callback_(createRequest<RequestT>(query));
+    zenohc::QueryReplyOptions options;
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_APP_CUSTOM });  // Update encoding.
+      query.reply(this->topic_config_.name, reply, options);
+    } else {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
+      query.reply(this->topic_config_.name, serdes::serialize(reply), options);
+    }
+  };
+
+  queryable_ = expectAsUniquePtr(
+      session_->zenoh_session.declare_queryable(topic_config_.name, { std::move(query), []() {} }));
+}
+
+template <typename RequestT, typename ReplyT>
+auto callService(const SessionPtr& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::chrono::milliseconds& timeout) -> std::optional<ReplyT> {
+  checkTemplatedTypes<RequestT, ReplyT>();
+
+  DLOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
+  bool done = false;
+  ReplyT reply_message;
+
+  auto on_reply = [&reply_message, &m, &topic_config](zenohc::Reply&& reply) {
+    const auto result = std::move(reply).get();
+    if (const auto& error = std::get_if<zenohc::ErrorMessage>(&result)) {
+      LOG(ERROR) << fmt::format("Error on {} reply: {}", topic_config.name, error->as_string_view());
+      return;
+    }
+
+    const auto& sample = std::get_if<zenohc::Sample>(&result);
+    DLOG(INFO) << fmt::format("Received answer of '{}'", sample->get_keyexpr().as_string_view());
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      std::unique_lock<std::mutex> lock(m);
+      reply_message = sample->get_payload().as_string_view();
+    } else {
+      auto payload = sample->get_payload();
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+      std::unique_lock<std::mutex> lock(m);
+      serdes::deserialize(buffer, reply_message);
```

Comment:
we would need to find a way to check that the type match. At least check that the encoding is not text. Ideally we would want to send the type as well.

Reply:
Added a check for the encoding.